### PR TITLE
28673 quick access new product fix

### LINKED ISSUE #28673: BO - Feature flag is enabled - Add new product from the quick access is NOK
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add screenshots

Previously reported here: https://github.com/PrestaShop/PrestaShop/pull/28594#pullrequestreview-993977071

The Add new product from the quick access is NOK when `feature flag is enabled`

### Expected behavior

Opens the modal of the product type when the user clicks on 'Add new product'. 
When the type is selected, the user is redirected to the product page with product type selected.

### Steps to reproduce

**Second issue**

**_Case1_**
1. Go to BO > Advanced Parameters > New & Experimental Features page
2. Enable `New product page - Single store` and save
3. In the Quick access > click on New product
4. See exception 

image

I attached a screen record

https://user-images.githubusercontent.com/16067358/171688808-581155bd-6fe2-4584-a76c-5474ca1094fe.mp4

**_Case2_**

1. Enable the multistore feature
2. Go to BO > Advanced Parameters > New & Experimental Features page
3. Enable `New product page - Multi store` and save
4. In the Quick access > click on New product
5. See exception 

### PrestaShop version(s) where the bug happened

develop

### PHP version(s) where the bug happened

7.2, 8.1

### If your bug is related to a module, specify its name and its version

_No response_

### PR DESCRIPTION
**Pull request by margud**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 8.0.x
| Description?      | Quick access fix with new product v2 page. Need to check new and old controllers.
| Type?             | bug fix 
| Category?         | BO
| BC breaks?        |  no
| Deprecations?     | no
| Fixed ticket?     | Fixes #28673 .
| How to test?      | PFirst you need to upgrade default and new theme. Then enable new product page feature and try to create new product using quick access. Same cases should be tested when multi-shop is enabled.
| Possible impacts? | Updated BO JS and CSS so it might have some impact.

#### Screenshots

Old controller: 
image
New controller: 
image

### CODE REVIEW

File: classes/controller/AdminController.php [hunk 1]
```diff
@@ -573,9 +574,18 @@ public function __construct($forceControllerName = '', $default_theme_name = 'de
             $this->can_import = true;
         }
 
+        if ($this instanceof AdminSearchConfController) {
```
Comment (margud) [inline classes/controller/AdminController.php]: This controller is called after original controller was called and it doesn't have correct metadata so it overrides correct information

**Review by NeOMakinG — DISMISSED** [review]
Why not, I guess this is the best compromise!

**Comment by NeOMakinG** [comment]
Tests are failing btw!

File: admin-dev/themes/default/scss/font.scss [hunk 2]
```diff
@@ -33,3 +33,7 @@
   /* Support for IE. */
   font-feature-settings: "liga";
 }
+
+.d-none {
+  display: none !important;
```
Comment (jolelievre) [inline admin-dev/themes/default/scss/font.scss]: The `!important` is mandatory in this case?
I'm surprised this class didn't exist already in Bootstrap 3 🤷‍♂️ But yeah turns out it's been introduced in Bootstrap 4

File: classes/QuickAccess.php [hunk 3]
```diff
@@ -135,4 +146,18 @@ public function toggleNewWindow()
 
         return $this->update(false);
     }
+
+    /**
+     * @return bool
+     */
+    private static function shouldRedirectToV2(): bool
```
Comment (jolelievre) [inline classes/QuickAccess.php]:
```suggestion
    private static function productPageV2Enabled(): bool
```

File: src/PrestaShopBundle/Twig/Extension/JsRouterMetadataExtension.php [hunk 4]
```diff
@@ -85,7 +85,7 @@ public function getJsRouterMetadata()
     {
         return [
             // base url for javascript router
-            'base_url' => $this->requestStack->getCurrentRequest()->getBaseUrl(),
+            'base_url' => $this->requestStack->getCurrentRequest() ? $this->requestStack->getCurrentRequest()->getBaseUrl() : '',
```
Comment (jolelievre) [inline src/PrestaShopBundle/Twig/Extension/JsRouterMetadataExtension.php]: Is this fix needed now that you handle this in the AdminController? Maybe it's a remnant of a previous implementation that didn't work as expected?

File: tools/assets/build.sh [hunk 5]
```diff
@@ -49,8 +49,8 @@ build_asset() {
       build "$PROJECT_PATH/themes/classic/_dev"
     ;;
     all)
-      build_asset admin-default
       build_asset admin-new-theme
+      build_asset admin-default
```
Comment (jolelievre) [inline tools/assets/build.sh]: Not sure it changes much things but why not
Unless this is now mandatory because the create product bundle must be built before the default theme? Since bundles are generated separately and included in the layout it shouldn't matter right?

**Review by jolelievre — COMMENTED** [review]
Thanks @margud just one renaming suggestion

The rest are just questions to be sure I understood everything

File: tools/assets/build.sh (hunk 5 above)
Comment (margud) [reply tools/assets/build.sh]: yeah, because default theme uses new theme JS it failed if the new theme was not setup before it. So thats why I changed it.

File: src/PrestaShopBundle/Twig/Extension/JsRouterMetadataExtension.php (hunk 4 above)
Comment (margud) [reply src/PrestaShopBundle/Twig/Extension/JsRouterMetadataExtension.php]: Well i no longer use this so i guess its not needed but still there is possibility that getCurrentRequest() will return null and if that heppens the shop gets exception if we try to call getBaseUri. For example when I tried using this on some cases shop would throw exception. But if you think its better to leave the old way i can change it back

File: admin-dev/themes/default/scss/font.scss (hunk 2 above)
Comment (margud) [reply admin-dev/themes/default/scss/font.scss]: Well I m not sure how much important is needed but this is how it looks on Bootstrap 4, so I added the same part.

File: src/PrestaShopBundle/Twig/Extension/JsRouterMetadataExtension.php (hunk 4 above)
Comment (jolelievre) [reply src/PrestaShopBundle/Twig/Extension/JsRouterMetadataExtension.php]: If we don't really need to change it then I prefer to leave it as it was indeed, as much as possible we try to limit the modification of the PR to only what's required this avoids introducing unexpected bugs or side effects

Besides even if it throws an exception it means something is going wrong, and having an exception doesn't hide the problem whereas a silent failing code could induce bugs that would be hard to identify then So I don't mind exceptions, I prefer them actually 😅

File: tools/assets/build.sh (hunk 5 above)
Comment (jolelievre) [reply tools/assets/build.sh]: Hmm, this one actually worries me a lot, we're not supposed to create dependencies between the two themes, at least on compilation time Which part of the new theme is required in the default one? I thought the create modal bundle was independent?

It's one thing to have a common bundle that is used in both themes, it's not ideal but we don't have another choice in this case, but the compilation process shouldn't be impacted by this 🤔

**Review by jolelievre — COMMENTED** [review]
Thanks @margud 

One thing to revert, and something bothers me about the theme compilation

File: admin-dev/themes/default/scss/font.scss [hunk 6]
```diff
@@ -33,3 +33,7 @@
...
+}
```
Comment (jolelievre) [inline admin-dev/themes/default/scss/font.scss]:
```suggestion
```
Since you now have a dedicated CSS bundle for the modal component you don't need to modify the default them anymore The less changes we perform in the default theme for this feature the better it is You can include this extra class in `create_product_default_theme.scss` instead. However, to limit its scope and side effects you should limit its use inside a target container based on an ID or class like:

```scss
#modal-create-product {
  .d-none {
    display: none !important;
  }
}
```

This way you guarantee that it works only in the expected scope of the modal and not anywhere else, thus limiting the possible side effects even in the default theme

File: admin-dev/themes/default/scss/partials/_header.scss [hunk 7]
```diff
@@ -930,3 +930,8 @@ $search-border-color: #bbcdd2;
     font-size: 22px;
   }
 }
+
+.modal-header .close {
+  float: right;
+  margin: -1.5rem 0 -1.25rem auto;
+}
```
Comment (jolelievre) [inline admin-dev/themes/default/scss/partials/_header.scss]:
```suggestion
```

Same suggestion as previous file, this should be handled in the dedicated SCSS file to limit the impact of this modification

File: admin-dev/themes/new-theme/.webpack/common.js [hunk 8]
```diff
@@ -136,6 +136,8 @@ module.exports = {
     webservice: './js/pages/webservice',
     zone: './js/pages/zone',
     country: './js/pages/country',
+    create_product: './js/pages/product/components/create-product',
+    new_product_modal: './scss/pages/product/new_product_modal.scss',
```
Comment (jolelievre) [inline admin-dev/themes/new-theme/.webpack/common.js]:
```suggestion
    create_product_default_theme: './scss/pages/product/create_product_default_theme.scss',
```

File: admin-dev/themes/default/scss/admin-theme.scss [hunk 9]
```diff
@@ -111,3 +111,7 @@
 
 //Print support
 @import "partials/print";
+
+//modal loader
+@import "../../new-theme/public/new_product_modal";
```
Comment (jolelievre) [inline admin-dev/themes/default/scss/admin-theme.scss]: I don't think you need this in the new theme At least for now all I see in the `new_product_modal.scss` file are dependencies on bootstrap or UI kit which should already be integrated in the new theme anyway

If you integrated special CSS for the component that would be different but for now there is none, although you'll see in my other comments that it could be enriched but it would still contain some duplicate imports. Even importing product stylesheet directly here can be improved. I would suggest the following structure:

1 - create a `create_product.scss` file containing the common stylesheet useful for both default and new theme (actually this file already exist https://github.com/PrestaShop/PrestaShop/blob/8.0.x/admin-dev/themes/new-theme/scss/pages/product/_create_product_modal.scss#L6-L5)
2 - in `admin-dev/themes/default/scss/partials/_product.scss` import this file, this is how we will update the new theme to handle the modal correctly on any new-theme page (since this CSS is the common one for the whole theme)
3 - create a `create_product_default_theme.scss` file, this one will be used as the entry endpoint for the generated bundle, so it can include the required bootstrap dependencies and of course the `create_product.scss` file as well, this will be used in default theme page so this bundle should be included automatically via the `setMedia` method in default theme pages (as you already did)

Ideally, the default bundle could be "protected" inside a designated scope (see my comment on font.css) to limit the impact, I don't know if it would be possible to also encapsulate the bootstrap dependencies inside a specific ID selector, this way the rules from bootstrap4 wouldn't impact those from bootstrap3 from the default theme

```scss
# That's the idea, maybe this ID selector is too limiting but maybe we can use another container to widen the scope a little, but trying to keep it as precise as possible to limit the side effects
#modal-create-product {
   @import "~bootstrap/scss/mixins/border-radius";
   @import "../../config/settings";
   @import "~prestakit/scss/spinners";
   @import "../../components/modal";
}
```

By following this structure we clearly separate what is inherent to the modal itself and needs to be used in both themes, from the required dependencies that are only needed in the default theme (because the new theme already has them), this way when one day we remove the default theme we can simply remove the `create_product_default_theme.scss` file since it won't be needed anymore, and the new theme will remain unchanged.

File: classes/controller/AdminController.php [hunk 10]
```diff
@@ -2614,6 +2614,7 @@ public function setMedia($isNewTheme = false)
 
             $this->addCSS(__PS_BASE_URI__ . $this->admin_webpath . '/themes/new-theme/public/theme.css', 'all', 0);
             $this->addJS(__PS_BASE_URI__ . $this->admin_webpath . '/themes/new-theme/public/main.bundle.js');
+            $this->addJS(__PS_BASE_URI__ . $this->admin_webpath . '/themes/new-theme/public/create_product.bundle.js');
```
Comment (jolelievre) [inline classes/controller/AdminController.php]: You should add the dedicated CSS for default theme here `create_product_default_theme.bundle.css`

File: admin-dev/themes/new-theme/js/pages/product/components/create-product.ts [hunk 11]
```diff
@@ -0,0 +1,30 @@
+/**
```
Comment (jolelievre) [inline admin-dev/themes/new-theme/js/pages/product/components/create-product.ts]: This file should be moved into the parent folder product I think, because it's not really a component but more an entry point to build a dedicated bundle (since it's a small one it doesn't need a dedicated folder like `edit`)

File: classes/controller/AdminController.php [hunk 12]
```diff
@@ -2698,6 +2701,20 @@ public function setMedia($isNewTheme = false)
             ],
         ]);
 
+        $username = $this->get('prestashop.user_provider')->getUsername();
```
Comment (jolelievre) [inline classes/controller/AdminController.php:2718]: I believe this part is only needed for the default theme no? since in new theme it is already handled via the twig extension. So maybe you can move this code inside the `else` block dedicated to the default theme?

File: admin-dev/themes/new-theme/scss/pages/product/create_product_default_theme.scss [hunk 13]
```diff
@@ -0,0 +1,17 @@
+#modal-create-product {
+  //modal loader
+  @import "~bootstrap/scss/mixins/border-radius";
+  @import "../../config/settings";
+  @import "~prestakit/scss/spinners";
+  @import "../../components/modal";
+  @import "create_product_modal";
+
+  .d-none {
+    display: none !important;
```
Comment (jolelievre) [inline admin-dev/themes/new-theme/scss/pages/product/create_product_default_theme.scss:12]: Seems like this line is bothering the scss linter because of the `!important`, so you need to ass an ignore comment exceptionally

```suggestion
    /* stylelint-disable declaration-no-important */
    display: none !important;
    /* stylelint-enable declaration-no-important */
```

**Review by jolelievre — COMMENTED** [review]
Hi @margud 

just some last few things to fix but it looks good! Also I tested the PR and I noticed an integration bug when the modal is loading (see in the video):

https://watch.screencastify.com/v/JbCI0B9GLFyYVeJqhXrC

It looks like an issue with the loading container that has an unwanted padding or margin, you should be able to fix this in `create_product_default_theme.scss` To debug it easily I believe you can force the loader to stay visible by commenting the code in this method:

https://github.com/PrestaShop/PrestaShop/blob/4762e95390ab98a2bf83293e80dfdaf3ed250a53/admin-dev/themes/new-theme/js/components/modal/iframe-modal.ts#L260-L264

(IframeModal is the base component of the create modal)

File: classes/controller/AdminController.php [hunk 14]
```diff
@@ -2660,6 +2663,20 @@ public function setMedia($isNewTheme = false)
...
             $this->addCSS(__PS_BASE_URI__ . $this->admin_webpath . '/themes/' . $this->bo_theme . '/css/overrides.css', 'all', PHP_INT_MAX);
+
+            $username = $this->get('prestashop.user_provider')->getUsername();
+            $token = $this->get('security.csrf.token_manager')
+                ->getToken($username)
+                ->getValue();
+
+            $this->context->smarty->assign([
+                'context_mode' => $this->context->mode,
+                'can_import' => $this->can_import,
```
Comment (jolelievre) [inline classes/controller/AdminController.php]: I don't see the relevance of these two fields why did you add them? Are they somehow used in a dependency we don't see in this PR? If they are not related to this PR they shouldn't be injected, we always modify the fewer we can to avoid unexpected side effects.

File: admin-dev/themes/new-theme/scss/components/_modal.scss [hunk 15]
```diff
@@ -56,12 +56,14 @@ $component-name: modal;
   }
 
   &-loader {
-    position: fixed;
+    position: absolute;
```
Comment (jolelievre) [inline admin-dev/themes/new-theme/scss/components/_modal.scss]: You can't change this fixed attribute because it was here for a good reason. This may fix your issue with the modal from quick access (I tested it works), but it creates a regression in other use cases. Especially in the specific price modal, here is the result with this modification:

https://watch.screencastify.com/v/G1Lr0kCT6FFJCUN1Wm2W

The scrollbars are broken, the loader can be moved when we use the mouse wheel. That's why it was forced as fixed to correctly overlay the whole content during the loading phase, here is the current expected behaviour without this modif:

https://watch.screencastify.com/v/79lgK0AFo16YaEocIpkX

Of course you couldn't know it, I should have added a comment to prevent future modifications, if you can add it please it will prevent other people to try and modify this in the future.

Comment (jolelievre) [reply admin-dev/themes/new-theme/scss/components/_modal.scss]:
```suggestion
    /* The fixed position is very important here, it allows keeping a cleaning loader over the modal and iframe content regardless of its size, especially for long iframe content which use scrollbars */
    position: fixed;
```

Comment (jolelievre) [reply admin-dev/themes/new-theme/scss/components/_modal.scss]: One last detail, in default theme the close button is not correctly placed while the modal is still loading, see this video:

https://watch.screencastify.com/v/pgVJnyfVPqNpJo6rLNlT

Once the modal is loaded it's ok, but it should be fine even during the initial load, and ideally if it can look like the same as the modal buttons in new theme it would be event better. This one is optional but since you might need to customize the CSS to customize this button to fix the small bug it might be a good occasion to make it prettier at the same time 😉.

**Review by jolelievre — CHANGES_REQUESTED** [review]
Thanks @margud 

Almost all good, I tested the modifications and there are still a few bugs related to the CSS. Sorry to bother you with this, but if it's not me the QA would probably have found it anyway 😅 (except the specific price modal bug which they may not have discovered now but probably later as a regression).

File: classes/controller/AdminController.php [hunk 16]
```diff
@@ -2698,6 +2713,18 @@ public function setMedia($isNewTheme = false)
...
+            ->getToken($username)
+            ->getValue();
+
+        $this->context->smarty->assign([
+            'js_router_metadata' => [
+                'base_url' => __PS_BASE_URI__ . basename(_PS_ADMIN_DIR_),
+                'token' => $token,
+            ],
+        ]);
+
```
Comment (jolelievre) [inline classes/controller/AdminController.php:2729]:
```suggestion
```

Comment (jolelievre) [reply classes/controller/AdminController.php:2729]: You moved this part in the default theme, but I think you forgot to clean it here

File: admin-dev/themes/new-theme/js/pages/product/components/create-product-modal.ts [hunk 17]
```diff
@@ -42,7 +42,7 @@ export default class CreateProductModal {
 
       const iframeModal = new FormIframeModal({
         id: 'modal-create-product',
-        formSelector: 'form[name="product"]',
+        formSelector: 'form[name="create_product"]',
```
Comment (jolelievre) [inline admin-dev/themes/new-theme/js/pages/product/components/create-product-modal.ts]: I'm sorry it should have been done from the start when this code was first created, but can you please move this selector into the ProductMap? Using `ProductMap.create.form`

Optionally the modal ID could also be placed in the ProductMap

Comment (jolelievre) [reply admin-dev/themes/new-theme/js/pages/product/components/create-product-modal.ts]: However I just tested the PR, it seems like the redirection after the submit doesn't work anymore I'm thinking maybe it's because of this modified selector The data returned by the `onFormLoaded` method are based on this form selector.

The problem here is that the name of the creation form and the name of the edition form are different so the selector can't work for both, so you're gonna need a selector that can work on both (not based on name). You could add a common `product-form` class on both and then use `form.product-form` as a common selector.

File: admin-dev/themes/new-theme/scss/pages/product/create_product_default_theme.scss [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+  @import "../../components/modal";
+  @import "create_product_modal";
+
+  .d-none {
+    /* stylelint-disable declaration-no-important */
+    display: none !important;
+    /* stylelint-enable declaration-no-important */
+  }
+
+  .modal-header {
```
Comment (jolelievre) [inline admin-dev/themes/new-theme/scss/pages/product/create_product_default_theme.scss:23]: I think we could customize this one a bit more to make it look like the one on the new theme:

```suggestion
  .modal-dialog {
    top: 50%;
    -webkit-transform: translateY(-50%);
    transform: translateY(-50%);
    width: 90%;
    max-width: 810px;
  }

  .modal-header {
      border-bottom: none;
```

**Review by jolelievre — COMMENTED** [review]
Thanks @margud 

almost there, but I tested the PR there seems to be a bug related to the form selector that changed (that's because the creation form name had to be changed in another PR). I suggested a solution in the comments

**Review by zuk3975 — DISMISSED** [review]

File: classes/controller/AdminController.php [hunk 19]
```diff
@@ -2629,6 +2630,7 @@ public function setMedia($isNewTheme = false)
             $this->addCSS(__PS_BASE_URI__ . $this->admin_webpath . '/themes/' . $this->bo_theme . '/css/' . $this->bo_css, 'all', 0);
             $this->addCSS(__PS_BASE_URI__ . $this->admin_webpath . '/themes/' . $this->bo_theme . '/css/vendor/titatoggle-min.css', 'all', 0);
             $this->addCSS(__PS_BASE_URI__ . $this->admin_webpath . '/themes/' . $this->bo_theme . '/public/theme.css', 'all', 0);
+            $this->addCSS(__PS_BASE_URI__ . $this->admin_webpath . '/themes/new-theme/public/create_product_default_theme.css', 'all', 0);
```
Comment (jolelievre) [inline classes/controller/AdminController.php]: Actually, it seems like this css needs to be added in both default AND new-theme.

I tested the modal on the product page and it displays fine because the product CSS has details for this modal, it also works fine in a legacy page thanks to this `create_product_default_theme.css`. However, when you test it on a migrated page it doesn't have the product CSS nor the modal CSS and it is displayed way too wide:

Capture d’écran 2022-09-07 à 19 51 47

So you should move this `addCss` call outside of the if/else block so that it's added on every page.

**Review by jolelievre — COMMENTED** [review]
Thanks @margud 

I just noticed one bug left and we should be good 🙂

**Comment by margud** [comment]
> Thanks @margud
> 
> almost there, but I tested the PR there seems to be a bug related to the form selector that changed (that's because the creation form name had to be changed in another PR). I suggested a solution in the comments

Its updated, please check it when you can. @jolelievre

File: classes/controller/AdminController.php (hunk 10 above)
Comment (jolelievre) [inline classes/controller/AdminController.php]: If this script is added in both themes, then this code should probably be added only once but outside of the `if/else` statement Also you should add a comment:

`// This is handled as an external common dependency for both themes, but once new-theme is the only one it should be integrated directly into the main.bundle.js file`

This will help future developers once we can get rid of the default theme

**Review by jolelievre — COMMENTED** [review]
Hi @margud 

This seems ok now (only one suggestion but optional), but can you please rebase the PR and target develop instead? We won't have time to include this in the 8.0 version.
